### DISCUSSION TITLE
Date filter is taking the app start date into account rather than the current OS date

### DISCUSSION THREAD

**Discussion by imdeaconu** [mikro-orm/mikro-orm#4077] [Q&A] [answered]
**Describe the bug**
Having a filter that's using new Date() to get the current date and time (useful when you need to have scheduled posts) isn't working, because it always takes the date the app started as a reference point instead of the OS time. I have to restart my app daily to make sure content that should be published isn't being hidden.

**To Reproduce**
1. Create a filter using new Date()
2. Create an entry, then change the date on your PC to a further date 
3. Create a new entry, then run .find() - you'll notice the latest entry isn't showing up

Code repro:

```
@Entity()
@Filter({
  name: "published",
  cond: {
    publishedAt: { $lte: new Date() },
  },
  default: true,
})
export class Song {
  @PrimaryKey({ type: "uuid", defaultRaw: "uuid_generate_v4()" })
  id!: string;

  @Property()
  title: string;

  @Property({ type: DateType })
  publishedAt?: Date;

  constructor(song: any) {
    this.id = song.id;
    this.title = song.title;
    this.publishedAt = new Date();
  }
}

const main = async () => {
  console.log("App start time: ", new Date());
  const orm = await MikroORM.init();
  const em = orm.em.fork();
  const song1 = em.create(Song, {
    title: "Anti-Hero",
  });
  await em.flush();

  setTimeout(async () => {
    console.log("Code ran after clock change");
    console.log("New time: ", new Date());

    const song2 = em.create(Song, {
      title: "Some other song",
    });
    await em.flush();
    const results = await em.find(Song, {});
    console.log(results);
  }, 60000);
};

main();

```
Console logs:
```
App start time:  2023-02-27T13:36:35.436Z
[info] MikroORM version: 5.6.12
[discovery] ORM entity discovery started, using ReflectMetadataProvider
[discovery] - processing entity Song
[discovery] - entity discovery finished, found 1 entities, took 21 ms
[info] MikroORM successfully connected to database mikro-bug on postgresql://postgres:*****@127.0.0.1:5432
[query] select 1 from pg_database where datname = 'mikro-bug' [took 2 ms]
[query] begin
[query] insert into "song" ("title", "published_at") values ('Anti-Hero', '2023-02-27') returning "id" [took 9 ms]
[query] commit
Code ran after clock change
New time:  2023-05-27T13:37:32.411Z
[query] begin
[query] insert into "song" ("title", "published_at") values ('Some other song', '2023-05-27') returning "id" [took 3 ms]
[query] commit
[query] select "s0".* from "song" as "s0" where "s0"."published_at" <= '2023-02-27' [took 3 ms]
[
  Song {
    id: '930be2e1-2b25-4daa-a10c-aa1fb891cf9d',
    title: 'Anti-Hero',
    publishedAt: 2023-02-27T13:36:35.776Z
  }
]

```

**Additional context**
I also tried to use em.raw('now()') for the filter, following one of the examples showed in the docs, but I encountered another bug that made my app crash. I could also report that one if it helps.

**Versions**

| Dependency | Version |
| - | - |
| node | 18.14.2 |
| typescript | 4.9.5 |
| mikro-orm | 5.6.12 |
| postgres | 5.6.12 |

**Comment by B4nan** [comment, marked as answer]
That is how code execution works, hard to call this a bug. You need to use a callback if you want to delay the execution.

```
@Filter({
  name: "published",
  cond: () => ({
    publishedAt: { $lte: new Date() },
  }),
  default: true,
})
```

**Reply by imdeaconu** [reply]
Oh, sorry I missed that! Thank you for you explanation and patience :)

**Reply by B4nan** [reply]
I'll be happy to look into the `em.raw()` issue if you can create a repro for that. The example from the docs is actually part of the ORM test suite (although just similar, not exactly the same), and the test is actually using postgres too, so it should work fine.

https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/filters/filters.postgres.test.ts#L65
